runtime/pkg/rduckdb/db.go

```diff
@@ -0,0 +1,1074 @@
+	// readMu controls access to readHandle
+	readMu ctxsync.RWMutex
+	// writeSem ensures only one write operation is allowed at a time
+	writeSem *semaphore.Weighted
+	// metaSem enures only one meta operation can run on a duckb handle.
+	// Meta operations are attach, detach, create view queries done on the db handle
+	metaSem *semaphore.Weighted
+	// localDirty is set to true when a change is committed to the remote but not yet reflected in the local db
+	localDirty bool
+	// ticker to peroiodically check if local db is in sync with remote
+	ticker  *time.Ticker
+	catalog *catalog
+
+	logger *slog.Logger
+
+	// ctx and cancel to cancel background operations
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+var _ DB = &db{}
+
+func (d *db) Close() error {
+	// close background operations
+	d.cancel()
+	d.ticker.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	_ = d.writeSem.Acquire(ctx, 1)
+	defer d.writeSem.Release(1)
+
+	err := d.readMu.Lock(ctx)
+	if err != nil {
+		return err
+	}
+	defer d.readMu.Unlock()
+
+	err = d.dbHandle.Close()
+	d.dbHandle = nil
+	return err
+}
+
+func (d *db) AcquireReadConnection(ctx context.Context) (*sqlx.Conn, func() error, error) {
+	if err := d.readMu.RLock(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	// acquire a connection
+	snapshot, err := d.catalog.acquireSnapshot(ctx)
+	if err != nil {
+		d.readMu.RUnlock()
+		return nil, nil, err
+	}
+
+	conn, err := d.dbHandle.Connx(ctx)
+	if err != nil {
+		d.readMu.RUnlock()
+		return nil, nil, err
+	}
+
+	err = d.prepareSnapshot(ctx, conn, snapshot)
+	if err != nil {
+		_ = conn.Close()
+		d.readMu.RUnlock()
+		return nil, nil, err
+	}
+
+	release := func() error {
+		err = d.catalog.releaseSnapshot(ctx, snapshot)
+		err = errors.Join(err, conn.Close())
+		d.readMu.RUnlock()
+		return err
+	}
+	return conn, release, nil
+}
+
+func (d *db) CreateTableAsSelect(ctx context.Context, name, query string, opts *CreateTableOptions) error {
+	d.logger.Debug("create table", slog.String("name", name), slog.Bool("view", opts.View))
+	err := d.writeSem.Acquire(ctx, 1)
+	if err != nil {
+		return err
+	}
+	defer d.writeSem.Release(1)
+
+	// pull latest changes from remote
+	err = d.pullFromRemote(ctx)
+	if err != nil {
+		return err
+	}
+
+	// check if some older version exists
+	oldMeta, _ := d.catalog.tableMeta(ctx, name)
+	if oldMeta != nil {
+		d.logger.Debug("old version", slog.String("version", oldMeta.Version))
+	}
+
+	// create new version directory
+	newVersion := newVersion()
+	newMeta := &tableMeta{
+		Name:           name,
+		Version:        newVersion,
+		CreatedVersion: newVersion,
+	}
+	var dsn string
+	if opts.View {
+		dsn = ""
+		newMeta.SQL = query
+		// special handling to ensure that if a view is recreated with the same name and schema then any views on top of this view still works
+		if oldMeta != nil && oldMeta.Type == "VIEW" {
+			newMeta.CreatedVersion = oldMeta.CreatedVersion
+		}
```

Comment:
Good catch. I did not think about it. 
I can not find a good solution for this. Both of the following solutions have some issues :

1. Create a view in a separate DB file similar to tables and then create another view in the snapshot to read from this view. Since all views present at the time of table creation will still be present in the db file  the view query will succeed and then we will be able to recreate views in any order in snapshot.

The problem is that because the views are not recreated in underlying db file so the view will continue to exist with the older schema if any underlying view's schema changes. There can be inconsistencies where `viewa` reports some schema whereas a dependent view `viewb` reports some other schema till the reconciler recreates `viewb`.

2. Create all views in single DB file and recreate all views in snapshot to read from these views. This will ensure that changes in underlying schema immediately fails dependent views. This may be problematic from the POV of concurrent writers.

May be this approach can be taken. For concurrent writers we may be able to fix this with a single lock for all views (but that needs to be a distributed lock so more complexity).

Reply:
> so the view will continue to exist with the older schema if any underlying view's schema changes

Don't we already have this problem today? And is this not also going to be a problem with all views in a single DB? Since we run models in DAG order without a transaction, if one model updates a view, the dependent views will be broken until they are reconciled (but only broken if queried, not breaking DB opens).

In general, IMO if you change the schema of a table/view, then it's acceptable behavior that querying a downstream view that has not been updated leads to undefined outcomes. People who are very sensitive to this behavior can just avoid views.

> Create all views in single DB file

Is also acceptable and concurrent writers can be handled with optimistic concurrency. However it is more complex and it would mean we have a pretty different implementation for views than for tables, so it would be nice to avoid.

----------

I can also think of two other options:
1. Have a queue for all the views to create. If creating a view for the first entry fails with a reference error, then move it to the back of the queue. Keep consuming the queue like this until an entire pass over the queue has happened without a single view being created.
  - This would have O(n^2) complexity, which is not so bad for the kinds of use cases we have, especially since view creation is fast.
  - But it is kind of silly...
2. Use our `duckdbsql` package to analyze each view's `SELECT` statement for dependent views. Then do a topological sort and create the views in that order. Link to the function: https://github.com/rilldata/rill/blob/aebc0ea9ed9e759a1634b8096590779804e31792/runtime/pkg/duckdbsql/ast.go#L157.
  - A big problem here is that the package is kind of best-effort and is prone to false positives (says something is a view/table, but it's actually an alias) and false negatives (doesn't detect that something is a view/table).
  - You would also need to filter out table references and do the topological sort, which will be kind of complicated

Reply:
> Don't we already have this problem today? And is this not also going to be a problem with all views in a single DB?

No. The views are broken today. With this approach, the view will succeed till they are reconciled (broken if dependent table also change).
So it is kind of a messy experience that some views will succeed while others will fail.


> In general, IMO if you change the schema of a table/view, then it's acceptable behavior that querying a downstream view that has not been updated leads to undefined outcomes. People who are very sensitive to this behavior can just avoid views.

Yeah I think it makes sense. I can go ahead with this approach then. It seems like the easiest approach and also something most consistent with the tables implementation.

Reply:
As it turns out I overlooked an issue with approach 1 that I had suggested.

Going ahead with the queue based approach now. We will see if it turns out to be performance issue. I am doing a sort by `created_at` so it should further reduce the failures.
May be writing in the same DB can be considered. We already have slightly different handling of views.
Want to avoid parsing SQL given the complexity and edge cases involved.